sync/src/main/java/tech/pegasys/artemis/sync/PendingBlocks.java

```diff
@@ -131,22 +134,40 @@ void onSlot(final SlotEvent slotEvent) {
     currentSlot = slotEvent.getSlot();
     if (currentSlot.mod(historicalBlockTolerance).equals(UnsignedLong.ZERO)) {
       // Purge old blocks
-      pruneBlocks(isTooOld());
+      prune();
     }
   }
 
+  @Subscribe
+  void onFinalizedCheckpoint(final FinalizedCheckpointEvent finalizedCheckpointEvent) {
+    this.latestFinalizedSlot = finalizedCheckpointEvent.getFinalizedSlot();
+  }
+
+  @VisibleForTesting
+  void prune() {
+    pruneBlocks(this::isTooOld);
+  }
+
   private boolean shouldIgnoreBlock(final BeaconBlock block) {
-    return isTooOld().test(block) || isFromFuture().test(block);
+    return isTooOld(block) || isFromFarFuture(block);
   }
 
-  private Predicate<BeaconBlock> isTooOld() {
-    final UnsignedLong slot = calculateBlockAgeLimit();
-    return (block) -> block.getSlot().compareTo(slot) <= 0;
+  private boolean isTooOld(final BeaconBlock block) {
+    return isFromAFinalizedSlot(block) || isOutsideOfHistoricalLimit(block);
   }
 
-  private Predicate<BeaconBlock> isFromFuture() {
+  private boolean isFromFarFuture(final BeaconBlock block) {
     final UnsignedLong slot = (calculateFutureBlockLimit());
-    return (block) -> block.getSlot().compareTo(slot) > 0;
+    return block.getSlot().compareTo(slot) > 0;
+  }
+
+  private boolean isOutsideOfHistoricalLimit(final BeaconBlock block) {
```

Comment:
Isn't this check redundant? How can a block be in the historical limit but not before a finalized slot?

Reply:
This check should only matter when we're out of sync.  It will prevent us from accumulating a huge set of pending blocks while we're doing our initial sync.